proposals/1533-host-specific-tasklist.md

```diff
@@ -0,0 +1,241 @@
+
+What CreateSession() and CompleteSession() really do is **scheduling a special activity** and get some information from the worker which executes the activity.
+
+* For CreateSession(), a special **session creation activity** will be scheduled on a global tasklist which is only used for this type of activity. During the execution of the activity, a signal containing the tasklist (depending on the session type, the tasklist will be either host specific or process specific) name will be sent back to the workflow. Then the tasklist name can be stored in the context for later use. The activity also performs heart beating throughout the whole lifetime of the session, so if the activity worker is down, the workflow can be notified and set the session state to "Failed". The CreateSession() call is considered as success when it gets the tasklist.
+
+* For RecreateSession(), the same thing happens. The only difference is that the session creation activity will be **scheduled on the resource specific task list instead of a global one**.
+
+* For CompleteSession(), another special **completion activity** will be scheduled on the process specific tasklist. The sole purpose of this activity is to stop the corresponding creation activity created by the CreateSession()/CreateSessionWithHostID() API call, so that the resource used by the session can be released. Note that, **the completion activity must be scheduled on the resource specific tasklist (not a global one)** since we need to make sure the completion activity and the creation activity it needs to stop are running by the same worker.
+
+### Activity Worker
+
+When EnableSessionWorker is set to true, two more activity workers will be started.
+
+* Session Creation Worker: this worker polls from a global tasklist whose name is derived from the tasklist name specified in the workerOption. This worker executes only one kind of activity: the special session creation activity. This special activity serves several purposes:
+
+   1. **Send the resource specific tasklist name** back to the workflow through signal.
+
+   2. Keep **heart beating** to cadence server till the end of the session.
+
+      If the heart beating returns EntityNotExist Error, this means the workflow/session has gone and this special activity should stop to release resource.
+
+      If on the other hand, the worker is down, cadence server will send heartbeat timeout error to the workflow, so the workflow can be notified and update the session state. Later ExecuteActivity() call with the failed session context will also fail.
```

Comment:
cadence server -> Cadence server
send heartbeat -> send a heartbeat
Later ExecuteActivity() call with -> Later, an ExecuteActivity() call to